[FLINK-24437][HA]Remove unhandled exception handler from CuratorFrame…

Aitozi opened the pull request:
…work before closing it

## What is the purpose of the change

This pull request enhances the usage of curator by removing unhandled exception handler from *CuratorFramework* before closing it 

## Verifying this change

This change is already covered by existing tests.

Aitozi commented:
Please help review this PR when you are free @tillrohrmann

Aitozi commented on flink-runtime/src/test/java/org/apache/flink/runtime/util/ZooKeeperUtilsTreeCacheTest.java:
```diff
@@ -65,14 +67,13 @@ public void setUp() throws Exception {
         configuration.set(
                 HighAvailabilityOptions.HA_ZOOKEEPER_QUORUM, testingServer.getConnectString());
 
-        client =
```
This line seems redundant, so  removed.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/ZooKeeperLeaderElectionTest.java:
```diff
@@ -434,27 +446,29 @@ public void testExceptionForwarding() throws Exception {
                                     configuration, NoOpFatalErrorHandler.INSTANCE));
 
-            doAnswer(invocation -> mockCreateBuilder).when(client).create();
+            doAnswer(invocation -> mockCreateBuilder).when(curatorResource.getClient()).create();
```
What this test effectively does is that it ensures that exception that are thrown in the `ZooKeeperLeaderElectionDriver.writeLeaderInformation()` will be forwarded to the fatal error handler.

Maybe we could actually get rid of the mocking by introducing a `CuratorFrameworkFacade` that we can configure to fail on the `create()` call or some other call, for example.

tillrohrmann reviewed (changes requested):
Thanks for creating this PR so quickly @Aitozi. The changes look good to me modulo some small suggestions concerning the naming. One thing we need to fix before merging this PR is to fix the failing test case: `ZooKeeperLeaderElectionTest.testExceptionForwarding:461 UnfinishedStubbing`.

Aitozi commented:
@tillrohrmann Thanks for your quickly reviewing :) , I have addressed your comments.
Regarding to the failure test, I have fixed it only by spying the correct *curatorFramework*  object, I have not come up with a good idea to introduce a `CuratorFrameworkFacade`, I will think about it later.

Aitozi commented:
About the naming of `CuratorFrameworkWithUnhandledErrorListener`, I have followed you suggestion but still have some doubts that with the name of `InternalCuratorResource` we can expand the class when there are some other resource shipped with curatorFramework client 🤔. 

Any way,  I agree for the naming for current version.
